contracts/stake/src/state.rs

```diff
@@ -214,21 +187,52 @@ impl StakeState {
         self.stakes.get_mut(&pk.to_bytes()).unwrap()
     }
 
-    /// Gets a mutable reference to the stake of a given key.
-    pub(crate) fn load_stake_mut(
-        &mut self,
-        pk: &PublicKey,
-    ) -> Option<&mut StakeData> {
-        self.stakes.get_mut(&pk.to_bytes())
-    }
-
     /// Rewards a `public_key` with the given `value`. If a stake does not exist
     /// in the map for the key one will be created.
     pub fn reward(&mut self, public_key: &PublicKey, value: u64) {
         let stake = self.load_or_create_stake_mut(public_key);
         stake.increase_reward(value);
     }
 
+    /// Total amount slashed from the genesis
+    pub fn slashed_amount(&self) -> u64 {
+        self.slashed_amount
+    }
+
+    /// Slash the given `to_subtract` amount from a `public_key` stake (if
+    /// any). Firstly the amount is subtracted from the reward, if that's
+    /// not enough the stake amount is touched.
+    pub fn slash(&mut self, public_key: &PublicKey, to_slash: u64) {
+        let slashed = self.get_stake_mut(public_key).map(|stake| {
+            if to_slash < stake.reward {
+                stake.reward -= to_slash;
+                to_slash
+            } else {
+                let slash_remaining = to_slash - stake.reward;
+                let mut slashed = stake.reward;
+                stake.reward = 0;
+
+                if let Some((stake_amt, _)) = stake.amount.as_mut() {
+                    let to_slash = min(*stake_amt, slash_remaining);
+                    *stake_amt -= to_slash;
+
+                    slashed += to_slash;
+
+                    let _: bool = rusk_abi::call(
+                        TRANSFER_CONTRACT,
+                        "sub_module_balance",
+                        &(STAKE_CONTRACT, to_slash),
+                    )
+                    .expect("Subtracting balance should succeed");
+                };
+                slashed
+            }
+        });
+        if let Some(slashed) = slashed {
+            self.slashed_amount += slashed
+        }
+    }
```

Comment:
Here's a suggestion for making this a little simpler - and it solves a bug where you subtract too much from the contract's balance under certain conditions.

```suggestion
    /// Slash the given `to_slash` amount from a `public_key` stake (if any).
    /// Firstly the amount is subtracted from the reward, if that's not enough
    /// the stake amount is touched.
    pub fn slash(&mut self, public_key: &PublicKey, to_slash: u64) {
        let stake = self
            .get_stake_mut(public_key)
            .expect("The stake to slash should exist");

        let mut slash_remaining = to_slash;

        // Slash the reward first
        let old_reward = stake.reward;
        stake.reward = stake.reward.saturating_sub(slash_remaining);
        slash_remaining -= old_reward - stake.reward;

        // Slash the stake amount, if it exists
        if let Some((amount, _)) = stake.amount.as_mut() {
            let old_amount = *amount;
            *amount = amount.saturating_sub(slash_remaining);
            slash_remaining -= old_amount - *amount;
        }

        // Subtract only the slashed amount from the balance of the module
        let slashed = to_slash - slash_remaining;

        let _: bool = rusk_abi::call(
            TRANSFER_CONTRACT,
            "sub_module_balance",
            &(STAKE_CONTRACT, slashed),
        )
        .expect("Subtracting balance should succeed");

        self.slashed_amount += slashed;
    }

```

Reply:
My intention was to avoid panicking while slashing.

However, if panic is allowed what do you think about this? At least to me seems more straightforward than your suggestion 

```rust
    /// Slash the given `to_subtract` amount from a `public_key` stake (if
    /// any). Firstly the amount is subtracted from the reward, if that's
    /// not enough the stake amount is touched.
    pub fn slash(&mut self, public_key: &PublicKey, to_slash: u64) {
        let stake = self
            .get_stake_mut(public_key)
            .expect("The stake to slash should exist");

        let staker_funds = stake.reward + stake.amount.unwrap_or_default().0;

        // Cannot slash more than the staker funds
        let to_slash = min(to_slash, staker_funds);

        if to_slash <= stake.reward {
            stake.reward -= to_slash;
        } else {
            // Deplete reward and update `to_slash` with the remaining amount to
            // slash from the stake amount.
            let remaining_slash = to_slash - stake.reward;
            stake.reward = 0;
            let (stake_amt, _) = stake
                .amount
                .as_mut()
                .expect("Trying to slash more than slashable_amount");

            *stake_amt -= remaining_slash;

            // Update the module balance to reflect the change in the amount
            // withdrawable from the contract
            let _: bool = rusk_abi::call(
                TRANSFER_CONTRACT,
                "sub_module_balance",
                &(STAKE_CONTRACT, remaining_slash),
            )
            .expect("Subtracting balance should succeed");
        }

        // Update the total slashed amount
        self.slashed_amount += to_slash;
    }
```